homework_1.py

```diff
@@ -0,0 +1,119 @@
+from pydantic import BaseModel
+from abc import ABC, abstractmethod
+
+
+class LibraryModel(BaseModel):
+    title: str
+    author: str
+    date: int
+
+
+class AbstractPrint(ABC):
+
+    @abstractmethod
+    def show_print_info(self):
+        raise NotImplementedError
+
+
+class Book(AbstractPrint):
+
+    def __init__(self, model: LibraryModel):
+        self._model = model
+
+    def __str__(self):
+        return f'{self._model.title} by {self._model.author} published on {self._model.date}'
+
+    def show_print_info(self):
+        return self.__str__()
+
+
+class Journal(Book):
+    pass
+
+
+def log_add_book(func):
+    def wrapper(self, book):
+        print(f'Added book: {book._model.title} by {book._model.author}.')
+        return func(self, book)
+    return wrapper
+
+
+def check_book_existence(func):
+    def wrapper(self, book):
+        if book in self.books:
+            return func(self, book)
+        else:
+            print(f'Book not found!!!')
+    return wrapper
+
+
+class Library:
+    def __init__(self):
+        self.books = []
+
+    @log_add_book
+    def add_book(self, book: Book):
+        self.books.append(book)
+        # print(f'Added book: {book._model.title} by {book._model.author}.')
+
+    @check_book_existence
+    def remove_book(self, book: Book):
+        if book in self.books:
+            self.books.remove(book)
+            print(f'Removed book: {book._model.title} by {book._model.author}')
+        else:
+            print(f'Book not found!')
+
+    def __iter__(self):
+        self.book_zero = 0
```

Comment:
Why it's called book_zero? This is counter, so it will be not only 0. Better to call it like count or counter

Reply:
Will fix. Thank you